src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs

```diff
@@ -0,0 +1,285 @@
+    public int MaxMatchLength { get; }
+
+    /// <summary>
+    /// The regular expression that matches the pattern near a signature.
+    /// <summary>
+    /// <remarks>
+    /// WARNING: This is distinct from the general regexes for this pattern that
+    /// can be applied to find all matches in an arbitrary input. This regex is
+    /// not run on the entire input but only on a substring determined by <see
+    /// cref="SignaturePrefixLength"/> and <see cref="MaxMatchLength"/>.
+    ///
+    /// Example: Let S be a signature character, X be a character--xpected to
+    /// match before the signature, Y be a character expected to match after the
+    /// signature, and Z be an optional character at the end of the signature.
+    ///
+    /// The regex must also be anchored to the beginning of the input.
+    ///
+    /// The regex may assume that the signature is present at the expected
+    /// location. This can allow regexes to be shared that would differ only by
+    /// signature. </remarks
+    public Regex ScopedRegex { get; }
+
+    /// <summary>
+    /// A rendering of the signature for fast, allocation-free lookup.
+    /// </summary>
+    /// <remarks>
+    /// This is implemented by packing the 3-4 byte ASCII bytes of the signature
+    /// into an integer, with the leftmost signature character in the least
+    /// signficant byte. However, this encoding should be considered an
+    /// implementation detail that can change.
+    /// </remarks>
+    public int PackedSignature { get; }
+
+    /// <summary>
+    /// Gets the pre-computed pattern for the given signature or null if there is none.
+    /// </summary>
+    public static CompiledHighPerformancePattern? ForSignature(string signature)
+    {
+        return s_patternsBySignature.TryGetValue(signature, out CompiledHighPerformancePattern? pattern) ? pattern : null;
+    }
+
+    /// <summary>
+    /// Regex options used for all high-performance regexes.
+    /// </summary>
+    /// <remarks>
+    /// We don't use 'NonBacktracking' here because it is not supported by the
+    /// regex source generator. However these high-performance patterns should
+    /// not cause problematic backtracking. They are extremely simple, anchored
+    /// to the front with at most one optional group at the end. Generated code
+    /// has been inspected to see that the only "loops" have at most 1
+    /// iteration. If more complex regexes are added, the generated code should
+    /// be re-examined.
+    ///
+    /// 'SingleLine' is used to ensure that '.' matches any character including
+    /// new lines. This is used to share regexes that can purely skip over the
+    /// signature that has already been matched before the regex is applied.
+    ///
+    /// Also note that the high-performance scanner does not allow substitution
+    /// of the regex engine and the regex engine built-in to .NET is always
+    /// used. Profiling showed no performance gain from RE2, even on .NET
```

Comment:
>Profiling showed no performance gain from RE2 [](http://example.com/codeflow?start=14&length=45)

we know this statement is suspect due to bugs in our impl but will revisit later

Reply:
Yes, more details in #158. Since this comment also says "this merits more scrutiny" and we are doing that as we, I don't think we need to change the comment.